h2o-core/src/main/java/water/ClientDisconnectCheckThread.java

```diff
@@ -0,0 +1,43 @@
+package water;
+
+import water.util.Log;
+
+class ClientDisconnectCheckThread extends Thread {
+
+  public ClientDisconnectCheckThread() {
+    super("ClientDisconnectCheckThread");
+    setDaemon(true);
+  }
+
+  private boolean isTimeoutExceeded(H2ONode client, long timeout) {
+    return (System.currentTimeMillis() - client._last_heard_from) >= timeout;
+  }
+
+  /**
+   * This method checks whether the client is disconnected from this node due to some problem such as client or network
+   * is unreachable.
+   */
+  private void handleClientDisconnect(H2ONode node) {
+    if(node != H2O.SELF) {
+      Log.warn("Client " + node + " disconnected!");
```

Comment:
This code is similar but not the same as UDPClientEvent.DISCONNECT - what is the reason for the difference?

Reply:
`UDPClientEvent.DISCONNECT` is the event which is triggered on purpose, usually, the clients decide to disconnect. In that case we also need to make sure that if this client is the watchdog client, we need to stop the cluster.

These cases are not handled here on purpose as this class is to handle client disconnections due to timeouts etc. It is up to the WatchDogClientExtension to handle correspondingly on the event that the watchdog client disappeared.